draft-rescorla-tls-esni.md

```diff
@@ -601,8 +601,52 @@ lie. In this design, the attacker would need to be able to find a
 Z which would expand into a key that would validly AEAD-encrypt
 a message of his choice, which should be intractable (Hand-waving alert!).
 
+# Alternative SNI Protection Designs
 
-# Alternate Encryption Design
+Alternative approaches to encrypted SNI may be implemented at the TLS or 
+application layer. In this section we describe several alternatives and discuss 
+drawbacks in comparison to the design in this document.
+
+## TLS-layer
+
+### TLS Tunnels
+
+In this variant, real TLS Client Hellos are tunneled within the early data payload 
+of an outer TLS connection established with the client-facing server. This requires 
+the client to have established a previous session — and obtained a PSK — with the 
+server. Problems with this approach are: (1) servers may not always be able to 
+distinguish inner Client Hellos from legitimate application data, (2) nested 0-RTT 
+data may not function correctly, (3) 0-RTT data may not be supported -- 
+especially under DoS -- leading to availability concerns, and (4) clients must bootstrap
+tunnels (sessions), costing an additional round trip and potentially revealing the SNI 
+during the initial connection. In contrast, encrypted SNI protects the SNI in a distinct 
+Client Hello extension and neither abuses early data nor requires a bootstrapping connection.
+
+### Combined Tickets
+
+In this variant, client-facing and hidden servers coordinate to produce "combined tickets" 
+that are consumable by both. Clients offer combined tickets to client-facing servers. 
+The latter parse them to determine the correct hidden server to which the Client Hello 
+should be forwarded. This approach is problematic due to non-trivial coordination between 
+client-facing and hidden servers for ticket construction and consumption. Moreover, 
+it requires a bootstrapping step similar to that of the previous variant. In contrast, 
+encrypted SNI requires no such coordination.
+
+## Application-layer
+
+### HTTP/2 CERTIFICATE Frames 
+
+In this variant, clients request secondary certificates with CERTIFICATE_REQUEST HTTP/2 
+frames after TLS connection completion. In response, servers supply certificates via TLS 
+exported authenticators {{!I-D.ietf-tls-exported-authenticator}} in CERTIFICATE frames. 
+Clients use a generic SNI for the underlying client-facing server TLS connection. 
+Problems with this approach include: (1) one additional round trip before peer 
+authentication, and (2) non-trivial application-layer dependencies and interaction. 
+In contrast, encrypted SNI induces no additional round trip and operates below the 
+application layer.
```

Comment:
Doesn't the scheme also has the bootstrap problem: i.e. distributing the name of the client-facing server that corresponds to the name of the server that the client actually wants to connect to?

I can see that such binding can sometimes advertised eagerly using the CERTIFICATE frames, but I do not think that can be done for every host that the client might want to connect to.

Reply:
Depending on how it's done, yeah, bootstrapping is also an issue here. I'll mention that.